pkg/netutil/networkwalker.go

```
@@ -0,0 +1,67 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package netutil
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+)
+
+type Found struct {
+	Network      *networkConfig
+	Req          string // The raw request string. name, short ID, or long ID.
+	MatchIndex   int    // Begins with 0, up to MatchCount - 1.
+	MatchCount   int    // 1 on exact match. > 1 on ambiguous match. Never be <= 0.
+	UniqueImages int    // Number of unique images in all found images.
+}
+
+type OnFound func(ctx context.Context, found Found) error
+
+type NetworkWalker struct {
+	Client  *CNIEnv
+	OnFound OnFound
+}
+
+// Walk walks networks and calls w.OnFound .
+// Req is name, short ID, or long ID.
+// Returns the number of the found entries.
+func (w *NetworkWalker) Walk(ctx context.Context, req string) (int, error) {
+	longIDExp, _ := regexp.Compile(fmt.Sprintf("^sha256:%s.*", regexp.QuoteMeta(req)))
```

The best practice here maybe not be to ignore errors here.